### PR TITLE
Replace capstone by rabbitizer

### PR DESCRIPTION
**Pull request by EllipticEllipsis**
- Add rabbitizer as a subrepo
- Replace all capstone use by rabbitizer
- C++ify a lot of the code for better error-checking
- Add ASAN options to the makefile

Output should be almost identical, with two differences:
- Because of slight changes in the way register tracking is done, some differences appear in the register tracking (but no functional differences in the code).
- `if`s now have newlines after them to improve readability.

Thanks to @AngheloAlf for doing most of the work clearing up the mess after my first attempt at conversion.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -113,10 +124,20 @@ $(shell mkdir -p $(BUILT_BIN))
...
+endif
+ifeq ($(DETECTED_OS),linux)
+# For traceback
+$(RECOMP_ELF): LDFLAGS   += -Wl,-export-dynamic
+endif
+
 # Too many warnings, disable everything for now...
-$(RECOMP_ELF): WARNINGS  += -Wno-unused-variable -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-function
+$(RECOMP_ELF): WARNINGS  += -Wpedantic -Wno-shadow -Wno-unused-variable -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-implicit-fallthrough
 # Do we really need no-strict-aliasing?
```
Comment (hensldm) [inline Makefile:141]: Yes, yes we do.
(Can delete the comment now)

File: recomp.cpp [hunk 2]
```diff
@@ -1,26 +1,47 @@
...
-#include <inttypes.h>
+#include <cassert>
+#include <cstdint>
+#include <cstdlib>
+#include <cstring>
+#include <cinttypes>
 #include <unistd.h>
 
-#include <algorithm>
+//#include <algorithm>
```
Comment (hensldm) [inline recomp.cpp:8]: Dead cod

File: recomp.cpp [hunk 3]
```diff
@@ -43,29 +67,132 @@ struct Edge {
...
+        this->b_liveout = 0;
+        this->b_livein = 0;
+        this->f_livein = 0;
+        this->f_liveout = 0;
+    }
+
+    void patchInstruction(rabbitizer::InstrId::UniqueId instructionId) {
+        // if (instructionId != rabbitizer::InstrId::UniqueId::cpu_nop) {
+        //     assert(!this->patched);
+        // }
```
Comment (hensldm) [inline recomp.cpp:128]: Dead cod

File: recomp.cpp [hunk 4]
```diff
@@ -43,29 +67,132 @@ struct Edge {
...
+
+        this->patched = true;
+        RabbitizerInstruction* innerInstr = this->instruction.getCPtr();
+        innerInstr->uniqueId = (RabbitizerInstrId)(instructionId);
+        innerInstr->descriptor = &RabbitizerInstrDescriptor_Descriptors[innerInstr->uniqueId];
+    }
+
+    void patchAddress(rabbitizer::InstrId::UniqueId instructionId, uint32_t newAddress) {
+        // assert(!this->patched);
```
Comment (hensldm) [inline recomp.cpp:137]: Dead cod

File: recomp.cpp [hunk 5]
```diff
@@ -358,182 +468,204 @@ static map<uint32_t, Function>::iterator find_function(uint32_t addr) {
...
+        default:
+            if (insn.instruction.modifiesRt()) {
+                return insn.instruction.GetO32_rt();
+            } else if (insn.instruction.modifiesRd()) {
+                return insn.instruction.GetO32_rd();
+            }
+            break;
+    }
+
+    // assert(!"No destination registers");
```
Comment (hensldm) [inline recomp.cpp:490]: Dead cod kind of
Would prefer more of an english statement/Todo if something needs to be done here.

File: recomp.cpp [hunk 6]
```diff
@@ -358,182 +468,204 @@ static map<uint32_t, Function>::iterator find_function(uint32_t addr) {
...
+                }
+                continue;
 
-                            default:
-                                assert(0);
+            case rabbitizer::InstrId::UniqueId::cpu_lw:
+            case rabbitizer::InstrId::UniqueId::cpu_ld:
+            case rabbitizer::InstrId::UniqueId::cpu_addiu:
+
+            // case rabbitizer::InstrId::UniqueId::cpu_addu: // used in jump tables for offset
```
Comment (hensldm) [inline recomp.cpp:514]: Dead cod

File: recomp.cpp [hunk 7]
```diff
@@ -653,19 +792,21 @@ static void pass1(void) {
                             uint32_t jtbl_addr = insns[lw].linked_value;
 
                             if (is_pic) {
-                                insns[i - 1].id = MIPS_INS_NOP;
+                                insns[i - 1].patchInstruction(rabbitizer::InstrId::UniqueId::cpu_nop);
                             }
 
                             // printf("jump table at %08x, size %u\n", jtbl_addr, num_cases);
```
Comment (hensldm) [inline recomp.cpp:798]: Dead cod

File: recomp.cpp [hunk 8]
```diff
@@ -686,183 +827,199 @@ static void pass1(void) {
...
+            case rabbitizer::InstrId::UniqueId::cpu_ldl:
+            case rabbitizer::InstrId::UniqueId::cpu_ldr:
+            case rabbitizer::InstrId::UniqueId::cpu_lh:
+            case rabbitizer::InstrId::UniqueId::cpu_lhu:
+            case rabbitizer::InstrId::UniqueId::cpu_lw:
+            case rabbitizer::InstrId::UniqueId::cpu_lwu:
+            case rabbitizer::InstrId::UniqueId::cpu_ldc1:
+            case rabbitizer::InstrId::UniqueId::cpu_lwc1:
+            case rabbitizer::InstrId::UniqueId::cpu_lwc2:
+            // case rabbitizer::InstrId::UniqueId::cpu_lwc3: // Seems unlikely that this is used
```
Comment (hensldm) [inline recomp.cpp:913]: Dead cod

File: recomp.cpp [hunk 9]
```diff
@@ -686,183 +827,199 @@ static void pass1(void) {
...
+            case rabbitizer::InstrId::UniqueId::cpu_swc1:
+            case rabbitizer::InstrId::UniqueId::cpu_swc2:
+                // case rabbitizer::InstrId::UniqueId::cpu_swc3:
```
Comment (hensldm) [inline recomp.cpp:916]: Dead cod

File: recomp.cpp [hunk 10]
```diff
@@ -686,183 +827,199 @@ static void pass1(void) {
...
+                    int32_t mem_imm = insns[i].instruction.getProcessedImmediate();
+
+                    if (mem_rs == rabbitizer::Registers::Cpu::GprO32::GPR_O32_gp) {
+                        unsigned int got_entry = (mem_imm + gp_value_adj) / sizeof(unsigned int);
+
+                        if (got_entry >= got_locals.size()) {
+                            got_entry -= got_locals.size();
+                            if (got_entry < got_globals.size()) {
+                                assert(insn.instruction.getUniqueId() == rabbitizer::InstrId::UniqueId::cpu_lw);
+                                // printf("gp 0x%08x %s\n", mem_imm, got_globals[got_entry].name);
```
Comment (hensldm) [inline recomp.cpp:928]: Dead cod

File: recomp.cpp [hunk 11]
```diff
@@ -686,183 +827,199 @@ static void pass1(void) {
...
+
+                                unsigned int dest_vaddr = got_globals[got_entry];
+
+                                insns[i].is_global_got_memop = true;
+                                insns[i].linked_value = dest_vaddr;
+                                // insns[i].label = got_globals[got_entry].name;
```
Comment (hensldm) [inline recomp.cpp:934]: Dead cod

File: recomp.cpp [hunk 12]
```diff
@@ -686,183 +827,199 @@ static void pass1(void) {
...
+                // could be insn?
+                rabbitizer::Registers::Cpu::GprO32 rt = insns[i].instruction.GetO32_rt();
+                rabbitizer::Registers::Cpu::GprO32 rs = insns[i].instruction.GetO32_rs();
+                int32_t imm = insns[i].instruction.getProcessedImmediate();
+
+                if (rs == rabbitizer::Registers::Cpu::GprO32::GPR_O32_zero) { // becomes LI
+                    insns[i].lila_dst_reg = get_dest_reg(insns[i]);
+                    insns[i].patchInstruction(UniqueId_cpu_li);
+                    insns[i].patchImmediate(imm);
+                } else if (/*rt == rs &&*/ rt !=
```
Comment (hensldm) [inline recomp.cpp:958]: Dead cod

File: recomp.cpp [hunk 13]
```diff
@@ -686,183 +827,199 @@ static void pass1(void) {
...
+                break;
         }
 
-        if (insn.id == MIPS_INS_ADDU && insn.operands[0].reg == MIPS_REG_GP && insn.operands[1].reg == MIPS_REG_GP &&
-            insn.operands[2].reg == MIPS_REG_T9 && i >= 2) {
+        if ((insn.instruction.getUniqueId() == rabbitizer::InstrId::UniqueId::cpu_addu) &&
+            (insn.instruction.GetO32_rd() == rabbitizer::Registers::Cpu::GprO32::GPR_O32_gp) &&
+            (insn.instruction.GetO32_rs() == rabbitizer::Registers::Cpu::GprO32::GPR_O32_gp) &&
+            (insn.instruction.GetO32_rt() == rabbitizer::Registers::Cpu::GprO32::GPR_O32_t9) && i >= 2) {
             // state->function_entry_points.insert(vaddr + (i - 2) * 4);
```
Comment (hensldm) [inline recomp.cpp:987]: Dead cod

File: recomp.cpp [hunk 14]
```diff
@@ -1009,67 +1156,70 @@ static void pass3(void) {
...
                 break;
             }
 
-            case MIPS_INS_JAL: {
+            case rabbitizer::InstrId::UniqueId::cpu_jal: {
                 add_edge(i, i + 1);
 
-                uint32_t dest = (uint32_t)insn.operands[0].imm;
+                uint32_t dest = insn.getAddress();
+                // fprintf(stderr, "%ld: %X: %X\n", i, insn.instruction.getRaw(), dest);
```
Comment (hensldm) [inline recomp.cpp:1222]: Dead cod

File: recomp.cpp [hunk 15]
```diff
@@ -1102,275 +1252,334 @@ static void pass3(void) {
...
+        case rabbitizer::InstrId::UniqueId::cpu_addiu:
+        case rabbitizer::InstrId::UniqueId::cpu_andi:
+        case rabbitizer::InstrId::UniqueId::cpu_ori:
+        case rabbitizer::InstrId::UniqueId::cpu_lb:
+        case rabbitizer::InstrId::UniqueId::cpu_lbu:
+        case rabbitizer::InstrId::UniqueId::cpu_lh:
+        case rabbitizer::InstrId::UniqueId::cpu_lhu:
+        case rabbitizer::InstrId::UniqueId::cpu_lw:
+        case rabbitizer::InstrId::UniqueId::cpu_lwl:
+        // case rabbitizer::InstrId::UniqueId::cpu_lwr:
```
Comment (hensldm) [inline recomp.cpp:1305]: Dead cod

File: recomp.cpp [hunk 16]
```diff
@@ -1102,275 +1252,334 @@ static void pass3(void) {
...
+
+        case rabbitizer::InstrId::UniqueId::cpu_beq:
+        case rabbitizer::InstrId::UniqueId::cpu_beql:
+        case rabbitizer::InstrId::UniqueId::cpu_bne:
+        case rabbitizer::InstrId::UniqueId::cpu_bnel:
+        case rabbitizer::InstrId::UniqueId::cpu_sb:
+        case rabbitizer::InstrId::UniqueId::cpu_sh:
+        case rabbitizer::InstrId::UniqueId::cpu_sw:
+        case rabbitizer::InstrId::UniqueId::cpu_swl:
+        // case rabbitizer::InstrId::UniqueId::cpu_swr:
```
Comment (hensldm) [inline recomp.cpp:1364]: Dead cod

File: recomp.cpp [hunk 17]
```diff
@@ -1102,275 +1252,334 @@ static void pass3(void) {
...
-            return TYPE_1S;
+        case rabbitizer::InstrId::UniqueId::cpu_jalr:
+            return TYPE_S;
 
-        case MIPS_INS_JR:
-            if (i.jtbl_addr != 0) {
-                i.operands[0].reg = i.index_reg;
+        case rabbitizer::InstrId::UniqueId::cpu_jr:
+            if (insn.jtbl_addr != 0) {
+                // insn.instruction.word = RAB_INSTR_PACK_rs(insn.instruction.word, insn.index_reg);
```
Comment (hensldm) [inline recomp.cpp:1393]: Dead cod

File: recomp.cpp [hunk 18]
```diff
@@ -1102,275 +1252,334 @@ static void pass3(void) {
...
+
+        default:
+            break;
+    }
+    if (instr.hasOperandAlias(rabbitizer::OperandType::cpu_rs)) {
+        return map_reg(instr.GetO32_rs());
+    } else if (instr.hasOperandAlias(rabbitizer::OperandType::cpu_rt)) {
+        return map_reg(instr.GetO32_rt());
+    } else {
+        // assert(!"No source registers");
```
Comment (hensldm) [inline recomp.cpp:1440]: Dead cod
Same comment as with the other commented assert, would prefer an english comment along with TODO if something needs to be done.

File: recomp.cpp [hunk 19]
```diff
@@ -1102,275 +1252,334 @@ static void pass3(void) {
...
+        ret |= map_reg(instr.GetO32_rs());
+    }
+    if (instr.hasOperandAlias(rabbitizer::OperandType::cpu_rt) && !instr.modifiesRt()) {
+        ret |= map_reg(instr.GetO32_rt());
+    }
+    return ret;
+}
+
+void pass4(void) {
+    vector<uint32_t> q; // Why is this called q?
```
Comment (hensldm) [inline recomp.cpp:1469]: Todo

File: recomp.cpp [hunk 20]
```diff
@@ -1102,275 +1252,334 @@ static void pass3(void) {
...
-                string name;
-                bool is_extern_function = false;
+                string_view name;
                 size_t extern_function_id;
-                auto it = symbol_names.find(insns[idx - 1].operands[0].imm);
+                uint32_t address = insns[i - 1].getAddress();
+
+                // TODO: Can this only ever be a J-type instruction?
+                auto it = symbol_names.find(address);
+                // auto it = symbol_names.find(insns[i - 1].operands[0].imm);
```
Comment (hensldm) [inline recomp.cpp:1549]: Dead cod

File: recomp.cpp [hunk 21]
```diff
@@ -1746,211 +1992,431 @@ static const char* dr(uint32_t reg) {
...
-}
 
-static void dump_cond_branch_likely(int i, const char* lhs, const char* op, const char* rhs) {
-    uint32_t target = text_vaddr + (i + 2) * 4;
-    dump_cond_branch(i, lhs, op, rhs);
-    if (!TRACE) {
-        printf("else goto L%x;\n", target);
+        if (!name.empty()) {
+            // printf("printf(\"%s %%x\\n\", %s);\n", name.c_str(),
+            // r((int)rabbitizer::Registers::Cpu::GprO32::GPR_O32_a0));
```
Comment (hensldm) [inline recomp.cpp:2195]: Dead cod

File: recomp.cpp [hunk 22]
```diff
@@ -1746,211 +1992,431 @@ static const char* dr(uint32_t reg) {
...
+
+        if (f.nret == 1) {
+            printf("v0 = ");
+        } else if (f.nret == 2) {
+            printf("temp64 = ");
+        }
+
+        if (!name.empty()) {
+            // printf("printf(\"%s %%x\\n\", %s);\n", string(name).c_str(),
+            // r((int)rabbitizer::Registers::Cpu::GprO32::GPR_O32_a0));
```
Comment (hensldm) [inline recomp.cpp:2208]: Dead cod

File: recomp.cpp [hunk 23]
```diff
@@ -1746,211 +1992,431 @@ static const char* dr(uint32_t reg) {
...
-                }
-                break;
 
-            case TYPE_1S_POS1:
-                if (!(insn.f_livein & map_reg(insn.operands[1].reg))) {
-                    printf("// fdead %llx ", (unsigned long long)insn.f_livein);
-                }
-                break;
+    // printf("// %llx, %llx, %llx, %llx : %d : %s\n", insn.b_livein, insn.b_liveout, insn.f_livein, insn.f_liveout,
+    //        insn_to_type(insn), insn.instruction.getOpcodeName().c_str());
```
Comment (hensldm) [inline recomp.cpp:2251]: Dead cod

File: recomp.cpp [hunk 24]
```diff
@@ -1959,374 +2425,234 @@ static void dump_instr(int i) {
...
                 printf("switch (%s) {\n", r(insn.index_reg));
 
                 for (uint32_t i = 0; i < insn.num_cases; i++) {
-                    uint32_t dest_addr = read_u32_be(rodata_section + jtbl_pos + i * 4) + gp_value;
+                    uint32_t dest_addr = read_u32_be(rodata_section + jtbl_pos + i * sizeof(uint32_t)) + gp_value;
                     printf("case %u: goto L%x;\n", i, dest_addr);
                     label_addresses.insert(dest_addr);
                 }
 
                 printf("}\n");
```
Comment (hensldm) [inline recomp.cpp:2648]: Dead cod

File: recomp.cpp [hunk 25]
```diff
@@ -2343,280 +2669,319 @@ static void dump_instr(int i) {
...
+            imm = insn.getImmediate();
 
-        case MIPS_INS_MFHI:
-            printf("%s = hi;\n", r(insn.operands[0].reg));
+            printf("%s = 0x%x;\n", r((int)insn.lila_dst_reg), imm);
+            // if (imm > 0) {
+            //     printf("%s = 0x%x;\n", r((int)insn.lila_dst_reg), imm);
+            // } else {
+            //     printf("%s = %i;\n", r((int)insn.lila_dst_reg), imm);
+            // }
```
Comment (hensldm) [inline recomp.cpp:2756]: Dead cod

File: recomp.cpp [hunk 26]
```diff
@@ -2911,6 +3279,7 @@ static void dump_c(void) {
             if (label_addresses.count(vaddr)) {
                 printf("L%x:\n", vaddr);
             }
+            // printf("// %s:\n", insn.disassemble().c_str());
```
Comment (hensldm) [inline recomp.cpp:3282]: Dead cod

File: recomp.cpp [hunk 27]
```diff
@@ -3268,6 +3637,7 @@ static void parse_elf(const uint8_t* data, size_t file_len) {
 
         // out_range.common_start = common_start;
         // out_range.common_order.swap(common_order);
```
Comment (hensldm) [inline recomp.cpp:3639]: Dead cod

File: recomp.cpp [hunk 28]
```diff
@@ -3397,5 +3844,6 @@ int main(int argc, char* argv[]) {
     // dump();
```
Comment (hensldm) [inline recomp.cpp:3844]: Dead cod

**Review by hensldm — COMMENTED** [review]
Nice work :) . 
Mostly focused on annotating dead code. I'm fine with commented dead code, but they really should have an English comment accompanying them which explains what is, why it is there, and why it is commented possibly with a TODO as needed.

File: recomp.cpp (hunk 16 above)
Comment (AngheloAlf) [reply recomp.cpp:1364]: I think there was a reason why this is commented out...

File: recomp.cpp (hunk 26 above)
Comment (AngheloAlf) [reply recomp.cpp]: I wrapped this into a `#if DUMP_INSTRUCTIONS` so it can be used as a debugging measure in the future

File: recomp.cpp (hunk 24 above)
Comment (AngheloAlf) [reply recomp.cpp:2635]: I added a note about this block, hope it is better this way

File: recomp.cpp (hunk 5 above)
Comment (AngheloAlf) [reply recomp.cpp:494]: @EllipticEllipsis any suggestion what we should do here?

File: recomp.cpp (hunk 10 above)
Comment (EllipticEllipsis) [reply recomp.cpp]: Isn't this a debugging print?

File: recomp.cpp (hunk 16 above)
Comment (EllipticEllipsis) [reply recomp.cpp:1352]: These are commented out because they were commented out in the original iirc.

File: recomp.cpp (hunk 18 above)
Comment (EllipticEllipsis) [reply recomp.cpp:1427]: I think we can just remove these if we introduced them, we changed the outputs a bit.

File: recomp.cpp (hunk 21 above)
Comment (EllipticEllipsis) [reply recomp.cpp]: I have no memory of what this was for...

File: recomp.cpp (hunk 24 above)
Comment (EllipticEllipsis) [reply recomp.cpp:2628]: Yes, this is a serious TODO, probably next on the list after resolving UB.

File: recomp.cpp (hunk 26 above)
Comment (EllipticEllipsis) [reply recomp.cpp]: I think it would be worth doing that for the register tracking print too.

File: recomp.cpp (hunk 28 above)
Comment (EllipticEllipsis) [reply recomp.cpp]: As in the original, this is also debugging.

File: recomp.cpp [hunk 29]
```diff
@@ -3841,7 +3827,6 @@ int main(int argc, char* argv[]) {
     pass4();
     pass5();
     pass6();
-    // dump();
```
Comment (EllipticEllipsis) [inline recomp.cpp:3397]: Leave this in since the function's still there.

Comment (AngheloAlf) [reply recomp.cpp:3397]: man is your PR rofl

File: recomp.cpp (hunk 10 above)
Comment (AngheloAlf) [reply recomp.cpp]: It doesn't make sense anymore since `got_globals` is a vector of integers now

**Review by hensldm — APPROVED** [review]

**Pull request merged**